### PR TITLE
8324580: SIGFPE on THP initialization on kernels < 4.10

### PR DESCRIPTION
**Pull request by zzambers**
**Problem:**
When THP is enabled, JDK reads `/sys/kernel/mm/transparent_hugepage/hpage_pmd_size` file to [detect](https://github.com/openjdk/jdk/blob/96607df7f055a80d56ea4c19f3f4fcb32838b1f8/src/hotspot/os/linux/hugepages.cpp#L206) large page size. However this file only [appeared](https://github.com/torvalds/linux/commit/49920d28781dcced10cd30cb9a938e7d045a1c94) in kernel 4.10 and does not exist on kernel old kernels (such as 3.10 used by RHEL-7).

This results in detected large page size of 0B and crash, when `-XX:+UseTransparentHugePages` is used on old kernel.
```
gdb --args jdk-23+6/bin/java -XX:+UseTransparentHugePages -Xmx128m -Xlog:pagesize -version
...
[0.005s][info][pagesize] Static hugepage support:
[0.005s][info][pagesize]   hugepage size: 2M
[0.005s][info][pagesize]   hugepage size: 1G
[0.005s][info][pagesize]   default hugepage size: 2M
[0.005s][info][pagesize] Transparent hugepage (THP) support:
[0.005s][info][pagesize]   THP mode: always
[0.005s][info][pagesize]   THP pagesize: 0B
[0.005s][info][pagesize] Shared memory transparent hugepage (THP) support:
[0.005s][info][pagesize]   Shared memory THP mode: unknown
[0.005s][info][pagesize] JVM will attempt to prevent THPs in thread stacks.
[0.005s][info][pagesize] UseLargePages=1, UseTransparentHugePages=1
[0.005s][info][pagesize] Large page support enabled. Usable page sizes: 4k. Default large page size: 0B.

Program received signal SIGFPE, Arithmetic exception.
[Switching to Thread 0x7ffff7fc2700 (LWP 31385)]
0x00007ffff68eeb20 in lcm(unsigned long, unsigned long) () from /home/tester/jdk-23+6/lib/server/libjvm.so
(gdb) bt
#0  0x00007ffff68eeb20 in lcm(unsigned long, unsigned long) () from /home/tester/jdk-23+6/lib/server/libjvm.so
#1  0x00007ffff64a68f2 in Arguments::apply_ergo() () from /home/tester/jdk-23+6/lib/server/libjvm.so
#2  0x00007ffff700458f in Threads::create_vm(JavaVMInitArgs*, bool*) () from /home/tester/jdk-23+6/lib/server/libjvm.so
#3  0x00007ffff6a2373f in JNI_CreateJavaVM () from /home/tester/jdk-23+6/lib/server/libjvm.so
#4  0x00007ffff7fe704b in InitializeJVM (ifn=<synthetic pointer>, penv=0x7ffff7fc1ea8, pvm=0x7ffff7fc1ea0) at src/java.base/share/native/libjli/java.c:1550
#5  JavaMain (_args=<optimized out>) at src/java.base/share/native/libjli/java.c:491
#6  0x00007ffff7feb1c9 in ThreadJavaMain (args=<optimized out>) at src/java.base/unix/native/libjli/java_md.c:650
#7  0x00007ffff7bc6ea5 in start_thread (arg=0x7ffff7fc2700) at pthread_create.c:307
#8  0x00007ffff76ebb0d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324580](https://bugs.openjdk.org/browse/JDK-8324580): SIGFPE on THP initialization on kernels &lt; 4.10 (**Bug** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [529afb72](https://git.openjdk.org/jdk/pull/17545/files/529afb72271f847790eb424fc8e2fdbfa502a77e)
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17545/head:pull/17545` \
`$ git checkout pull/17545`

Update a local copy of the PR: \
`$ git checkout pull/17545` \
`$ git pull https://git.openjdk.org/jdk.git pull/17545/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17545`

View PR using the GUI difftool: \
`$ git pr show -t 17545`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17545.diff">https://git.openjdk.org/jdk/pull/17545.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17545#issuecomment-1907154220)

### CODE REVIEW

**Comment by tstuefe** [comment]
Hi @zzambers, thanks for finding this one.

I would not disable THPs in this case, though. I would set it to the default static huge page size, if that is available and <=, say, 16 MB.

Reasoning:
- Getting the THP size wrong is not so bad:
  - if its smaller than reality, we miss some THP clustering opportunities at the border of reservations, but no big deal
  - if its larger than reality, we may unnecessarily align up memory sizes and lose some memory that way. As long as the page size is not absurdly large (e.g. 1GB pages), its fine.

Therefore my proposal would be, in os_linux.cpp, when determining _large_pagesize, to:

if (thp_pagesize == 0) {
  set it to MIN2(static default large page size, 16MB);
}

**Comment by tstuefe** [comment]
(16 MB because it is the largest page size in a group of sizes I loosely consider "small large pages", and over-allocating by that is still okay)

**Comment by zzambers** [comment]
@tstuefe thank you for your feedback, I'll make suggested change and update this PR

**Comment by tstuefe** [comment]
> @tstuefe thank you for your feedback, I'll make suggested change and update this PR

Since we don't disable, and I usually prefer error symptoms from solutions in bug titles anyway, could you pls change the JBS bug description and the PR title to something like:

"SIGFPE on THP initialization on kernels < 4.10"
?

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -3874,8 +3874,16 @@ void os::Linux::large_page_init() {
...
-    assert(HugePages::thp_pagesize() > 0, "Missing OS info");
     _large_page_size = HugePages::thp_pagesize();
+    if (_large_page_size == 0) {
+        // Unknown THP page size => falback to default static hugepage size
+        if (!HugePages::supports_static_hugepages()) {
+            warn_no_large_pages_configured();
+            UseLargePages = false;
+            return;
+        }
+        _large_page_size = MIN2(HugePages::default_static_hugepage_size(), 16 * M);
```
[the remark this answers is not in the document]
Comment (zzambers) [reply src/hotspot/os/linux/os_linux.cpp]: I have updated this code and test. (just with inverted if, and fixed kernel version to 4.10)

File: src/hotspot/os/linux/os_linux.cpp [hunk 2]
```diff
@@ -3874,8 +3874,16 @@ void os::Linux::large_page_init() {
     // In THP mode:
     // - os::large_page_size() is the *THP page size*
     // - os::pagesizes() has two members, the THP page size and the system page size
-    assert(HugePages::thp_pagesize() > 0, "Missing OS info");
     _large_page_size = HugePages::thp_pagesize();
+    if (_large_page_size == 0) {
+        // Unknown THP page size => falback to default static hugepage size
```
[the remark this answers is not in the document]
Comment (zzambers) [reply src/hotspot/os/linux/os_linux.cpp]: Comments updated with resolution of other thread.

**Review by tstuefe — APPROVED** [review]
Good. You'll need a second reviewer, since this is hotspot.

**Comment by zzambers** [comment]
When testing this on older kernel I have ran into failure of `runtime/os/TestTracePageSizes.java` test:
```
java.lang.AssertionError: Page sizes mismatch: 4 != 2048
	at TestTracePageSizes.main(TestTracePageSizes.java:307)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
	at java.base/java.lang.Thread.run(Thread.java:1575)
```
As far as I can tell this is test only problem.

**Details:**
Test [relaxes](https://github.com/openjdk/jdk/blob/2cd1ba6a52eafffa65d0f2532a07fff89f9cea0e/test/hotspot/jtreg/runtime/os/TestTracePageSizes.java#L301) page size check if mapping uses THP. [Detection logic](https://github.com/openjdk/jdk/blob/2cd1ba6a52eafffa65d0f2532a07fff89f9cea0e/test/hotspot/jtreg/runtime/os/TestTracePageSizes.java#L368) depends of `THPeligible` line from smaps, when THP mode is `always` (default on RHEL-7). This information however only [appeared in kernel 5.0](https://github.com/torvalds/linux/commit/7635d9cbe8327e131a1d3d8517dc186c2796ce2e). All tests pass with THP mode switched to `madvise` on RHEL-7.

Test is already being skipped in some configurations. So It could probably be skipped also for `kernel < 5.0 && thp_mode == always`.

I can include fix in this PR.

**Comment by zzambers** [comment]
> > I can include fix in this PR.
> 
> No, please open a new issue and put it on my name. Let's just ship this fix, we should get it out.

Ok, I'll keep this PR as it is and create another issue instead.

Thank you

**Comment by kstefanj** [comment]
Thanks for the ping @tstuefe =)

I think the approach to use the default large page size is good, but I think we should try to add the logic to `THPSupport::scan_os()` so it returns with a reasonable page size set. This will make the logging show the correct page size as well. If I'm not mistaken, right now the log will still render 0 as the THP page size:
```
[0.005s][info][pagesize] Transparent hugepage (THP) support:
[0.005s][info][pagesize]   THP mode: always
[0.005s][info][pagesize]   THP pagesize: 0B
```

If we update `THPSupport::scan_os()` to call `scan_default_hugepagesize()` if reading `hpage_pmd_size` fails we should be good. We can add a check there to limit the "default size" to 16M (which sounds reasonable) and maybe log that we did this limiting. I'm not certain what we should do if reading the default hugepage size fail (returns 0), but I'm leaning towards having 2M as a general fallback. It will only be used if THP mode is configured to `madvide` or `always` and in those cases we likely have at least a 2M page size (and I'm not sure this should ever happen). We could also add code to `validate_thps_configured()` to return false if the thp page size is 0.

An alternative to how we log this is to add a tag to the "THP pagesize" line, something like:
```
[0.005s][info][pagesize]   THP pagesize: 0B (configured/default/limited/fallback)
```
But I'm not sure how useful that is compared to have an additional log-line for the limited and fallback case.

If we go with this approach we should also update the testing code to do the same dance.

What do you think?

**Comment by tstuefe** [comment]
Hi @kstefanj ,

> Thanks for the ping @tstuefe =)
> 
> I think the approach to use the default large page size is good, but I think we should try to add the logic to `THPSupport::scan_os()` so it returns with a reasonable page size set. This will make the logging show the correct page size as well. If I'm not mistaken, right now the log will still render 0 as the THP page size:
> 
> ```
> [0.005s][info][pagesize] Transparent hugepage (THP) support:
> [0.005s][info][pagesize]   THP mode: always
> [0.005s][info][pagesize]   THP pagesize: 0B
> ```
> 

I'm unsure about this. I originally wanted the stuff in huge pages.cpp/hpp to not "lie" - as in not making assumptions. It was supposed to truthfully reflect the system's settings without making assumptions. The assumption-making should happen in the code above, in os_linux.cpp. 

That strict separation was born out of working with the code before, which had been a big ball of tangled yarn with respect to who calls what and who lies about what. But maybe I am too religious about it now. But whatever we decide, I would like the information that we *assume* a THP size instead of just reading it from the OS to be reflected somewhere in the log and printed settings.

> If we update `THPSupport::scan_os()` to call `scan_default_hugepagesize()` if reading `hpage_pmd_size` fails we should be good. We can add a check there to limit the "default size" to 16M (which sounds reasonable) and maybe log that we did this limiting. I'm not certain what we should do if reading the default hugepage size fail (returns 0), but I'm leaning towards having 2M as a general fallback. It will only be used if THP mode is configured to `madvide` or `always` and in those cases we likely have at least a 2M page size (and I'm not sure this should ever happen). We could also add code to `validate_thps_configured()` to return false if the thp page size is 0.

Sure, we can do that. Though I believe THPs are implemented atop of explicit huge pages, so I would assume them to be always available. 

> 
> An alternative to how we log this is to add a tag to the "THP pagesize" line, something like:
> 
> ```
> [0.005s][info][pagesize]   THP pagesize: 0B (configured/default/limited/fallback)
> ```
> 
> But I'm not sure how useful that is compared to have an additional log-line for the limited and fallback case.
> 
> If we go with this approach we should also update the testing code to do the same dance.
> 
> What do you think?

See above. I can do both. If we move THP pagesize recognition, including assumptions, down into hugepages.cpp (that was @zzambers first attempt too), I would like the information that we made an assumption reflected in the printout.

**Comment by kstefanj** [comment]
> I'm unsure about this. I originally wanted the stuff in huge pages.cpp/hpp to not "lie" - as in not making assumptions. It was supposed to truthfully reflect the system's settings without making assumptions. So, if information is missing, print 0. The assumption-making should happen in the code above, in os_linux.cpp.
> 
> That strict separation was born out of working with the code before, which had been a big ball of tangled yarn with respect to who calls what and who lies about what. But maybe I am too religious about it now. But whatever we decide, I would like the information that we _assume_ a THP size instead of just reading it from the OS to be reflected somewhere in the log and printed settings.
> 

I agree this is a good approach, and things have really improved in this area over the last few years. It still felt like this kind of logic was a better fit in `THPSupport::scan_os()` and reading the default and using that would still not violate the "truth aspect" too much. I do agree that doing the limiting and potentially using a fixed default is more towards "lying", and it would be possible to split that out to `os_linux.cpp`. Not sure if this is worth it, so making sure the log is clear would be my choice.

> > If we update `THPSupport::scan_os()` to call `scan_default_hugepagesize()` if reading `hpage_pmd_size` fails we should be good. We can add a check there to limit the "default size" to 16M (which sounds reasonable) and maybe log that we did this limiting. I'm not certain what we should do if reading the default hugepage size fail (returns 0), but I'm leaning towards having 2M as a general fallback. It will only be used if THP mode is configured to `madvide` or `always` and in those cases we likely have at least a 2M page size (and I'm not sure this should ever happen). We could also add code to `validate_thps_configured()` to return false if the thp page size is 0.
> 
> Sure, we can do that. Though I believe THPs are implemented atop of explicit huge pages, so I would assume them to be always available.
> 

I think so too, that's why I think having a "fixed default" if nothing can be read should be ok. If we manage to read that the setting is `always`or `madvise` we should try to use them even if we can't read a size of the pages. 

> > An alternative to how we log this is to add a tag to the "THP pagesize" line, something like:
> > ```
> > [0.005s][info][pagesize]   THP pagesize: 0B (configured/default/limited/fallback)
> > ```
> > What do you think?
> 
> See above. I can do both. If we move THP pagesize recognition, including assumptions, down into hugepages.cpp (that was @zzambers first attempt too), I would like the information that we made an assumption reflected in the printout.

So what would you prefer, some kind of tag like suggested above, or a messages saying what we do, like:
```
[0.005s][info][pagesize] THP page size not present in sysfs
[0.005s][info][pagesize] Using default huge page size for THP
```
Maybe those should be on debug level if going this route.

**Comment by tstuefe** [comment]
@zzambers I am fine with the current version of the patch too, but would ask you to:
- change, like @kstefanj suggested, the behaviour when no static hugepage was detected. In that case just use 2MB as an assumed default. I think that's usually right, and if it is wrong it is no big deal.
- add a log line somewhere that clearly states "we chose such and such page size as assumed THP page size"

**Comment by zzambers** [comment]
Ok, I'll do these fixes.

Btw. there is one more use of `HugePages::thp_pagesize` is in [THPStackMitigation](https://github.com/openjdk/jdk/blob/a9c6e87c6ac8f278e4e241a77aef3479f7b5f2e4/src/hotspot/os/linux/os_linux.cpp#L978). It currently only works for known thp_pagesize (>0). Should I change it to make use of `os::large_page_size` to also do "mitigation" with assumed page size?

**Review by kstefanj — COMMENTED** [review]
@zzambers, I'm ok with not moving the logic to keep the class true to the system conf. But I see one problem with using `os::large_page_size()` for `THPStackMitigation`. That mitigation is in play even if transparent huge pages are not enabled explicitly for the JVM (`-XX:+UseTransparentHugePages`), ie the THP kernel mode is `always`. I guess it is very unlikely, but a user could be setting `LargePageSizeInBytes=1g` to use 1g pages for the heap for example. This would set the default large page size to 1g and it would be returned by `os::large_page_size()`. We should not use that page size for the stack mitigation, since it will still be subject to the real "thp pagesize".  

For this reason it would have been good if `thp_pagesize()` returned a proper value, but again, I see the arguments for keeping it true to the configuration.

Since the mitigation haven't worked on old kernels up until now (since the THP pagesize returned is 0), I guess that could be handled as a separate issue.

**Comment by tstuefe** [comment]
> For this reason it would have been good if `thp_pagesize()` returned a proper value, but again, I see the arguments for keeping it true to the configuration.

Yeah, I am almost convinced now that modifying `HugePages::thp_pagesize()` would have been better. Maybe with an extra boolean flag that means "We did guess the THP page size".

> Btw. there is one more use of HugePages::thp_pagesize for [THPStackMitigation](https://github.com/openjdk/jdk/blob/a9c6e87c6ac8f278e4e241a77aef3479f7b5f2e4/src/hotspot/os/linux/os_linux.cpp#L978). It currently only does something for known thp_pagesize (>0). Should I change it to make use of os::large_page_size to also do "mitigation" based on assumed page size?

No, definitely not; for the reason @kstefanj mentioned. We don't want 1GB stacks if someone uses 1GB pages :=)

**Comment by zzambers** [comment]
> Yeah, I am almost convinced now that modifying `HugePages::thp_pagesize()` would have been better. Maybe with an extra boolean flag that means "We did guess the THP page size".

So far I have just modified logic. One more idea which come to my head was to just make fallback computation as method to HugePages (without logging). And than do something like this:
```
_large_page_size = HugePages::thp_pagesize();
if (_large_page_size == 0) {
    log_info(pagesize) ("Cannot determine THP page size (kernel < 4.10 ?)");
    ...
    _large_page_size = HugePages::thp_pagesize_fallback();
}
```

**Comment by tstuefe** [comment]
> > Yeah, I am almost convinced now that modifying `HugePages::thp_pagesize()` would have been better. Maybe with an extra boolean flag that means "We did guess the THP page size".
> 
> So far I have just modified logic. One more idea which come to my head was to just make fallback computation as method to HugePages (without logging). And then do something like this:
> 
> ```
> _large_page_size = HugePages::thp_pagesize();
> if (_large_page_size == 0) {
>     log_info(pagesize) ("Cannot determine THP page size (kernel < 4.10 ?)");
>     ...
>     _large_page_size = HugePages::thp_pagesize_fallback();
> }
> ```

Yes, we can do that too.

**Review by kstefanj — APPROVED** [review]
I'm fine with this, but moving the fallback calculation to the `HugePage` class also seems like a good idea. 

Maybe also file a enhancement to investigate if we should the fallback for THP mitigation as well.

**Review by kstefanj — APPROVED** [review]

**Comment by zzambers** [comment]
I have moved fallback computation logic to HugePages class. Hopefully this will be last iteration. :)

**Comment by kstefanj** [comment]
I think this is good, thanks for going the extra mile.

**Review by tstuefe — APPROVED** [review]
Okay!

Lets wait for GHAs to be green, then it can be integrated.

**Comment by zzambers** [comment]
Tests are good, if you are Ok with latest change, I'll proceed with integration.

File: test/hotspot/jtreg/runtime/os/HugePageConfiguration.java [hunk 3]
```diff
@@ -86,6 +86,18 @@ public long getThpPageSize() {
         return _thpPageSize;
     }
 
+    public long getThpPageSizeFallback() {
```
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/os/HugePageConfiguration.java:89]: Can you please slightly rename this to `getThpPageSizeOrFallback`, and write a comment along the lines of:
```
Returns the THP page size; if the kernel does not expose that, return a guessed page size.
Mimics HugePages::thp_pagesize_fallback() in hotspot and must be kept in sync with that function.
```

**Review by tstuefe — APPROVED** [review]
Still good. Small change request inline. I don't need another review, if you do the fixup, feel free to slash integrate.

**Review by kstefanj — APPROVED** [review]
Ship it :)

**Comment by zzambers** [comment]
Unfortunately I was informed that there is a build failure after this (clash with intermediate change). Fix is in progress:
https://github.com/openjdk/jdk/pull/17879

**Pull request closed without merging**